Add checks and explicit errors when cart buttons are missing data

tomalec opened the pull request:
### Changes proposed in this Pull Request:

Add checks and explicit errors when cart buttons are missing data
to help debugging issues with other extensions manipulating WC Core add & remove from cart buttons or events.

This would help identifying issues like the one reported at https://wordpress.org/support/topic/js-error-on-add-to-cart-when-logged-out/

### Checks:
<!-- Mark completed items with an [x] -->
* [x] Does your code follow the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Screenshots:

![image](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/17435/25994140-5e50-41f5-be62-4acd12a1c054)



### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Build the branch
2. open shop page
3. Run synthetic/malformed cart events
```js
jQuery( document.body ).trigger( 'added_to_cart' );
jQuery( document.body ).trigger( 'removed_from_cart' );
```
8. Check for errors in the console
6. Add an item to the cart using UI.
7. Remove all product ids `document.querySelectorAll('[data-product_id]').forEach((e)=>e.removeAttribute('data-product_id'))`. 8. Remove item from mini cart using UI
8. Check for errors in the console
9. Smoke test not tempered events and buttons for adding to & removing from cart, for regression


### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Tweak - Add checks and explicit errors when cart buttons are missing data.

tomalec committed:
Add checks and explicit errors when cart buttons are missing data

to help debugging issues with other extensions manipulating WC Core add & remove from cart buttons or events.

eason9487 commented on assets/js/src/integrations/classic.js:
```diff
@@ -60,14 +60,20 @@ export function classicTracking(
 		// Get product ID from data attribute (archive pages) or value (single product pages).
 		const productID = parseInt(
-			button[ 0 ].dataset.product_id || button[ 0 ].value
+			button?.[ 0 ]?.dataset.product_id || button?.[ 0 ].value
```
If the purpose is to help debugging, perhaps it also needs to consider the case where the `button` parameter is an empty array/jQuery collection.

```js
jQuery( document.body ).trigger( 'added_to_cart', [ , , jQuery('match-nothing') ] );
```

![image](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/17420811/1afdc017-bcf1-454a-935e-22775b005417)

eason9487 commented on assets/js/src/integrations/classic.js:
```diff
@@ -60,14 +60,20 @@ export function classicTracking(
+			button?.[ 0 ]?.dataset.product_id || button?.[ 0 ].value
 		);
 
+		if ( isNaN( productID ) ) {
```
Although `parseInt` won't return a type other than number or `NaN`, it would be better to avoid using `isNaN` when possible because it has some [confusing special-case behavior](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/isNaN#description) that leads to semantics ambiguity.

Maybe it could be replaced with `Number.isNaN`.

eason9487 reviewed (approved):
Thanks for the improvement. LGTM.

Commented on two minor concerns, and I believe it won't need another round even if there are related changes.

tomalec committed:
Consider cases where button is an empty jQuery collection

Address https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/443#discussion_r1651954213

Fix grammar in the error message.

tomalec committed:
Use `Number.isNaN` as a safer alternative to `isNaN`

Address https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/443#discussion_r1651961681

tomalec commented:
Thanks @eason9487 for the detailed review :)
